### PR TITLE
[5.2] Media field: selectable folders

### PR DESCRIPTION
**Pull request by Fedik**
### Summary of Changes

The PR add posibility to select directories usign the Media field.
This is helpfull when the site have multiple Media adapters.

### Testing Instructions

Run `npm install`.

Create custom field Media,
Pick the directory in the "Directory" option.

Edit TinyMCE (plugin config) "Images Directory" (when "Images Drag and Drop" is On). And test TinyMCE drag and drop upload.

### Actual result BEFORE applying this Pull Request

You able to pick only one of folder under `/images`

### Expected result AFTER applying this Pull Request

You can pick any directory that available in the Media manager

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: https://docs.joomla.org/Media_form_field_type
- [ ] No documentation changes for docs.joomla.org needed
- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/288
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Review by dgrammatiko — APPROVED** [review]

**Comment by brianteeman** [comment]
Works fine for custom fields but what about in the media plugin?

image

**Comment by dgrammatiko** [comment]
@Fedik you could probably patch also this event: https://github.com/joomla/joomla-cms/blob/dfb572de83d16a25e7eee673b3875e18750e688b/administrator/components/com_media/resources/scripts/components/tree/tree.vue#L85 
So selecting a folder on the left (disk/folder tree) could be also a valid way to select the folder

**Comment by Fedik** [comment]
@brianteeman I forgot it exists, should be good now.
@dgrammatiko updated

**Comment by brianteeman** [comment]
Is it possible to use something other than "types" as I am sure people will confuse it with "type"

**Comment by dgrammatiko** [comment]
> Is it possible to use something other than "types" as I am sure people will confuse it with "type"


That would be a b/c break so no

**Comment by Fedik** [comment]
Nope, unfortunately. It is what we already use https://github.com/joomla/joomla-cms/blob/bea594e404224ab404fe76ca55e5f9910227f183/libraries/src/Form/Field/MediaField.php#L237
For the Media field.

**Comment by brianteeman** [comment]
ok fair enough. I searched for "types =" and got no hits - didn't think about "types                                                    ="

;)

**Comment by Fedik** [comment]
I also updated TinyMCE upload directory for "drag and drop", to use this feature.

File: administrator/language/en-GB/plg_editors_tinymce.ini [hunk 1]
```diff
@@ -11,6 +11,7 @@ PLG_TINY_CORE_BUTTONS="CMS Content"
 PLG_TINY_DND_ADDITIONALDATA="Additional Data"
 PLG_TINY_DND_ALTTEXT="Image Description (Alt Text)"
 PLG_TINY_DND_EMPTY_ALT="No Description"
+PLG_TINY_DND_FILE_EXISTS_ERROR="File "%s" already exists."
```
Comment (Quy) [inline administrator/language/en-GB/plg_editors_tinymce.ini]:
```suggestion
PLG_TINY_DND_FILE_EXISTS_ERROR="File \"%s\" already exists."
```

**Comment by laoneo** [comment]
Nice work. I tested it with DPMedia and the FTP plugin, unfortunately when I select a folder from the FTP adapter, the field stays empty in the media custom field.

**Comment by dgrammatiko** [comment]
@laoneo is right, this approach will never work for anything else other than the local filesystem. Maybe having `dolder/path#adapterName://folder` could work, but would need patches all over I guess

**Comment by Fedik** [comment]
Thanks for checking. I did not have a time to test with external adapters, and expected it works.
I will update later

**Comment by laoneo** [comment]
There is a free version of DPMedia with the FTP adapter if you want to try it by yourself.

**Comment by Fedik** [comment]
I have checked. The field "Content - DPMedia" override configuration form for the Media Custom field (with own field for it).
You need to disable this plugin then it will work.

After I disable that plugin it works form me
Screenshot 2024-06-06_19-04-16

**Comment by laoneo** [comment]
I tested it, when it was disabled. The error I did, I didn't select the folder in the media manager, I just navigated to the folder I wanted to test and then hit the select button. Like in this screenshot:
image

**Comment by Fedik** [comment]
> I just navigated to the folder I wanted to test and then hit the select button

Hm, it also works for me.
You navigating with Tree select or Breadcrumbs? Breadcrumbs does not do anything, but Tree should work.

**Comment by laoneo** [comment]
I clicked the folder inside the main area, where the folders are listed, not the tree or breadcrumbs.

**Comment by Fedik** [comment]
> I clicked the folder inside the main area, where the folders are listed, not the tree or breadcrumbs.

So you clicked it twice to open?
The media manager cannot handle it as select.
 
There only one way to select something in media manager, is tick the checkbox on the image/folder.
And one addittional way for folders is to click the folder in Tree view.

Yeah, well, it can be confusing.

**Comment by laoneo** [comment]
I double clicked on the last folder inside the main area and then the select button. But I had to do it by being in the parent and then select the folder  with the checkbox and not navigating into the folder and then click select.

**Comment by Fedik** [comment]
I have update PR, now the folder should be also selectable on "double click".

**Comment by laoneo** [comment]
This worked perfectly, also with navigation into another folder and with other adapters than the local one.

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on 3223f82594c969c70c2c8bafbd39fcec9495d976<br /><br />See my last comment<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43579">issues.joomla.org/tracker/joomla-cms/43579</a>.</sub>

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on 3223f82594c969c70c2c8bafbd39fcec9495d976<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43579">issues.joomla.org/tracker/joomla-cms/43579</a>.</sub>

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on 3223f82594c969c70c2c8bafbd39fcec9495d976<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43579">issues.joomla.org/tracker/joomla-cms/43579</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43579">issues.joomla.org/tracker/joomla-cms/43579</a>.</sub>

**Comment by dgrammatiko** [comment]
Just a note on the person that would merge this PR: depending on the order of committing this PR and #43823 the one that would be second would have to update the extra event introduced in #43823.

**Comment by Hackwar** [comment]
This feature is nice and I will merge it as soon as drone has passed. One thing which I find strange however is, that you can set it to a subfolder, but then you can still move up the folders and select a different folder. This is also present in the current implementation, but if I set it to folder X, I wouldn't expect to be able to upload in a parent folder of that.

**Comment by Hackwar** [comment]
Thanks @Fedik for your contribution. :-)

**Pull request merged**